### PR TITLE
Do not cache access token if refresh token is set

### LINKED ISSUE #171: Critical problem with token_hash in Api
### General Information

- Mode (Sandbox/Live): Live
- PayPal-Debug-ID(s) (from any logs): Many
- Version of Python used: 3.4
- SDK Version: 1.9.0

### Issue Description

We are using PayPal for 'classic' payments using the redirect method and now in a new app using future payments. This mixture of services used is dangerous, since the API object is global and caches a value, that maybe shouldn't be cached.

This is the steps to reproduce:
1. Add a PayPal future payment method using the authorization code from the mobile device in `api.get_refresh_token(auth_code)`
1.1. This will call the `get_token_hash` method and pass on the `authorization_code`.
2. Create a Authorization/Payment (in the name of User A) using that `refresh_token` and a valid `correlation_id`.
2.2. This will call `get_token_hash` with the `refresh_token` of a user, and the Authorization/Payment will be done in his name
3. Create another Payment in the name of User B, without a `refresh_token`. This should create a 'classic' payment using a rediect to PayPal.
3.1. But the token_hash of the user before is still cached in the global api object. Through this caching the token_hash of the user A will be sent for the Payment of user B.
3.2. Somehow this leads to that the Payment of user B is instantly payed using the details of user A.

The only workaround of this is creating a new API object per Payment and passing the new API object into the `Payment.create` method, so no caching occurs.

### PR DESCRIPTION
**Pull request by jaypatel512**
- Fixes #171.

### CODE REVIEW

File: paypalrestsdk/api.py [hunk 1]
```diff
@@ -92,21 +92,24 @@ def get_token_hash(self, authorization_code=None, refresh_token=None, headers=No
 
         else:
             self.validate_token_hash()
-            if self.token_hash is not None:
+            if refresh_token is None && self.token_hash is not None;
                 return self.token_hash
-            else:
+            elif refresh_token is None:
```
Comment (bluk) [inline paypalrestsdk/api.py]: Do we need the `refresh_token is None` here and line 95 (would the we have executed line 90 instead in this method)? If we do need it, can the `if refresh_token is None` be extracted out one level and put the old code inside the new `if` block?

File: paypalrestsdk/api.py [hunk 2]
```diff
@@ -92,21 +92,24 @@ def get_token_hash(self, authorization_code=None, refresh_token=None, headers=No
...
                 "Authorization": ("Basic %s" % self.basic_auth()),
                 "Content-Type": "application/x-www-form-urlencoded",
                 "Accept": "application/json", "User-Agent": self.user_agent
             }, headers or {}))
-
-        return self.token_hash
+        if refresh_token is None:
+            self.token_hash = token;
+            return self.token_hash
+        else :
```
Comment (bluk) [inline paypalrestsdk/api.py]: Nitpick, extra space for the `else :`. We might also compress lines 110 to 112 to just be a `return token` if that reads better.

Comment (jaypatel512) [inline paypalrestsdk/api.py]: Done.

File: paypalrestsdk/api.py (hunk 1 above)
Comment (jaypatel512) [inline paypalrestsdk/api.py]: We need to check this after `self.validate_token_hash`.

File: paypalrestsdk/api.py [hunk 3]
```diff
@@ -92,21 +92,23 @@ def get_token_hash(self, authorization_code=None, refresh_token=None, headers=No
...
                 "Authorization": ("Basic %s" % self.basic_auth()),
                 "Content-Type": "application/x-www-form-urlencoded",
                 "Accept": "application/json", "User-Agent": self.user_agent
             }, headers or {}))
-
-        return self.token_hash
+        
+        if refresh_token is None:
+            self.token_hash = token;
+        return token;
```
Comment (danakatz) [inline paypalrestsdk/api.py]: should we remove the semicolons here for style consistency?

**Pull request merged**